monty-31/src/dft/backward.rs

```diff
@@ -108,40 +281,24 @@ impl<MP: MontyParameters + TwoAdicData> MontyField31<MP> {
         Self::backward_pass(a, &root_table[0]);
     }
 
-    #[inline(always)]
-    fn backward_64(a: &mut [Self], root_table: &[Vec<Self>]) {
-        assert_eq!(a.len(), 64);
-
-        // Safe because a.len() == 64
-        let (a0, a1) = unsafe { a.split_at_mut_unchecked(a.len() / 2) };
-        Self::backward_32(a0, &root_table[1..]);
-        Self::backward_32(a1, &root_table[1..]);
-
-        Self::backward_pass(a, &root_table[0]);
-    }
-
-    #[inline(always)]
-    fn backward_128(a: &mut [Self], root_table: &[Vec<Self>]) {
-        assert_eq!(a.len(), 128);
-
-        // Safe because a.len() == 128
-        let (a0, a1) = unsafe { a.split_at_mut_unchecked(a.len() / 2) };
-        Self::backward_64(a0, &root_table[1..]);
-        Self::backward_64(a1, &root_table[1..]);
-
-        Self::backward_pass(a, &root_table[0]);
-    }
+    /// Assumes `a.len() > 8`
+    #[inline]
+    fn backward_fft_recur(input: &mut [Self], root_table: &[Vec<Self>]) {
+        const ITERATIVE_FFT_THRESHOLD: usize = 1024;
 
-    #[inline(always)]
-    fn backward_256(a: &mut [Self], root_table: &[Vec<Self>]) {
-        assert_eq!(a.len(), 256);
+        let n = input.len();
+        if n <= ITERATIVE_FFT_THRESHOLD {
+            Self::backward_iterative(input, root_table);
+        } else {
+            assert_eq!(n, 1 << (root_table.len() + 1));
 
-        // Safe because a.len() == 256
-        let (a0, a1) = unsafe { a.split_at_mut_unchecked(a.len() / 2) };
-        Self::backward_128(a0, &root_table[1..]);
-        Self::backward_128(a1, &root_table[1..]);
+            // Safe because a.len() > ITERATIVE_FFT_THRESHOLD
+            let (a0, a1) = unsafe { input.split_at_mut_unchecked(n / 2) };
+            Self::backward_fft_recur(a0, &root_table[1..]);
+            Self::backward_fft_recur(a1, &root_table[1..]);
 
-        Self::backward_pass(a, &root_table[0]);
+            Self::backward_pass(input, &root_table[0]);
```

Comment:
If we reach this point, `input.len() > ITERATIVE_FFT_THRESHOLD` so instead of calling `backward_pass` why not just run:

`let packed_input = <Self as Field>::Packing::pack_slice_mut(input);`
`backward_iterative_layer_0(packed_input, &root_table[0]);`

Along the same idea, wouldn't it be cleaner for `backward_fft_recur` to accept a packed input: `packed_input: &mut [Self::Packed]` instead of the current: `input: &mut [Self]`?

As `backward_iterative` immediately converts `input` to `&mut [Self::Packed]` anyway (and we can call `backward_iterative_layer_0` instead of `backward_pass`) this should be an easy change.

This would also help if for some reason `pack_slice_mut` doesn't get optimized away.